### ISSUE TITLE
PermGen leak

### ISSUE THREAD

**Issue by patriknw** [akka/akka#17729]
When shutting down actor system.
Reported here: https://groups.google.com/d/msg/akka-user/XDVyWUT2u7E/o4jWaq9Le5cJ

**Comment by patriknw** [comment]
I have not confirmed the leak

**Comment by krchniam** [comment]
Sorry for the delay, I've created sample project: https://github.com/krchniam/akka-remote-bug-17729.

When deployed on GlassFish 3.1.2.2. After subsequent disabling application, there is one leaked thread from ExecutionContext.Implicits.global (two threads are used during shutdown, ForkJoinPool-3-worker-7 remains in state WAITING).

**Comment by patriknw** [comment]
thanks, is it possible to reproduce the problem without any external dependencies?

**Comment by krchniam** [comment]
I tried to cut dependencies to minimum (akka with all it`s dependencies and slf4j), so project runs on every JEE application server. My point is that, redeploying application with akka remoting on JEE server is causing PermGen leak (thread is holding WebAppClassLoader or EarLibClassLoader).

**Comment by patriknw** [comment]
ok, I was looking for something that could reproduce it without JEE app server.
Thanks anyway, we will look into it. You are welcome to submit a pull request if you know what is wrong.

**Comment by krchniam** [comment]
I'm not sure why is ```Remoting#shutdown``` using ```import scala.concurrent.ExecutionContext.Implicits.global``` instead of ```import system.dispatcher``` (if it is still running during remoting shutdown). That imports creates global ```ExecutionContext```. When akka is shut down, ```system.dispatcher``` thread pool will be shut down. But ```scala.concurrent.ExecutionContext.Implicits.global``` remains active with one thread holding ClassLoader.  My workaround shutting down ```scala.concurrent.ExecutionContext.Implicits.global``` via ```ActorSystem``` termination callback.

**Comment by drewhk** [comment]
I think it was about system.dispatcher not running at that point, but we need to revisit this. Using Implicits.global is definitely not the way to go. Thanks for reporting!

**Comment by patriknw** [comment]
@krchniam I have tried your sample with latest Glassfish 4.1, and I can confirm that number of live WebappClassLoader instances increase with one for each re-deployment and that used metaspace is increasing.

It is not solved by replacing the use of `Implicits.global`. I have replaced it with `system.dispatcher` in the `ShutdownAndFlush` usage in remoting. I verifed that the callbacks are actually running with the system.dispatcher. We should do that change anyway.

I also removed everything from the AkkaActorSystemEJB class, i.e. not starting ActorSystem and not using slf4j. WebappClassLoader instances increase in the same way when I re-deplay that thing, i.e. this is not an Akka bug as far as I can see. I will fix the Implicits.global anyway.

I used this for re-deployment:

    bin/asadmin deploy --force=true /Users/patrik/tmp/akka-remote-bug-17729/target/akka-remote-bug-17729-0.0.1-SNAPSHOT.war
